File: anagram.js
const stringOne = 'loop'
const stringTwo = 'pool'

console.log(`Is ${stringOne} an anagram of ${stringTwo}:`, isValidAnagram(stringOne, stringTwo))

function isValidAnagram(string1, string2) {
    let isValid = true
    let sortedOne = mergeSort(string1)
    let sortedTwo = mergeSort(string2)
    sortedOne.map((letter, i) => {
        if(letter != sortedTwo[i]) {
            isValid = false
        }
    })
    return isValid
}

function mergeSort(myString) {
    if(myString.length <= 1) {
        return myString
    }

    let result = []
    const mid = Math.ceil(myString.length / 2)
    const left = mergeSort(myString.substring(0, mid))
    const right = mergeSort(myString.substring(mid, myString.length))
    let leftIndex = 0
    let rightIndex = 0

    while(leftIndex < left.length && rightIndex < right.length) {
        if(left[leftIndex] < right[rightIndex]) {
            result.push(left[leftIndex])
            leftIndex++
        } else {
            result.push(right[rightIndex])
            rightIndex++
        }
    }

    while(leftIndex < left.length) {
        result.push(left[leftIndex])
        leftIndex++
    }

    while(rightIndex < right.length) {
        result.push(right[rightIndex])
        rightIndex++
    }

    return result
}
// [a, b, c] [d, e]
//  ^            ^
//  ^         ^

File: anagram.rs
fn main() {
    println!("{}", is_anagram(String::from("loop"), String::from("pool")));
}

fn is_anagram(string1: String, string2: String) -> bool {
    let mut is_valid = false;
    if merge_sort(string1) == merge_sort(string2) {
        is_valid = true;
    }
    return is_valid;
}

fn merge_sort(my_string: String) -> String {
    if my_string.len() <= 1 {
        return my_string;
    }
    let mut result = String::from("");
    let mid = my_string.len() / 2;
    let mut left = merge_sort(String::from(&my_string[..mid]));
    let mut right = merge_sort(String::from(&my_string[mid..]));
    let mut left_index = 0;
    let mut right_index = 0;

    while left_index < left.len() && right_index < right.len() {
        let left_element = left.chars().nth(left_index).unwrap();
        let right_element = right.chars().nth(right_index).unwrap();
        if left_element < right_element {
            result.push(left_element);
            left_index += 1;
        } else {
            result.push(right_element);
            right_index += 1;
        }
    }

    while left_index < left.len() {
        result.push(left.chars().nth(left_index).unwrap());
        left_index += 1;
    }

    while right_index < right.len() {
        result.push(right.chars().nth(right_index).unwrap());
        right_index += 1;
    }

    return result;
}
